lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,230 @@
+
+          # Attempt to fetch the license content
+          license_content = self.class.retrieve_license(match[1])
+          @nuspec_remote_license_file = Licensee::ProjectFiles::LicenseFile.new(license_content, { uri: match[1] }) if license_content
+        end
+
+        class << self
+          def strip_html(html)
+            return unless html
+
+            if html.downcase.include?("<html")
+              ReverseMarkdown.convert(html, unknown_tags: :bypass)
+            else
+              html
+            end
+          end
+
+          def ignored_url?(url)
+            # Many Microsoft packages that now use <license> use this for <licenseUrl>
+            # No need to fetch this page - it just contains NuGet documentation
+            url == "https://aka.ms/deprecateLicenseUrl"
+          end
+
+          def text_content_url(url)
+            # Convert github file URLs to raw URLs
+            return url unless match = url.match(/https?:\/\/(?:www\.)?github.com\/([^\/]+)\/([^\/]+)\/blob\/(.*)/i)
+            "https://github.com/#{match[1]}/#{match[2]}/raw/#{match[3]}"
+          end
+
+          def retrieve_license(url)
+            return unless url
+            return if ignored_url?(url)
+
+            # Transform URLs that are known to return HTML but have a corresponding text-based URL
+            text_url = text_content_url(url)
+
+            raw_content = fetch_content(text_url)
+            strip_html(raw_content)
+          end
+
+          def fetch_content(url, redirect_limit = 5)
+            url = URI.parse(url) if url.instance_of? String
+            return @response_by_url[url] if (@response_by_url ||= {}).key?(url)
+            return if redirect_limit == 0
+
+            begin
+              response = Net::HTTP.get_response(url)
+              case response
+              when Net::HTTPSuccess     then
+                @response_by_url[url] = response.body
+              when Net::HTTPRedirection then
+                redirect_url = URI.parse(response["location"])
+                if redirect_url.relative?
+                  redirect_url = url + redirect_url
+                end
+                # The redirect might be to a URL that requires transformation, i.e. a github file
+                redirect_url = text_content_url(redirect_url.to_s)
+                @response_by_url[url] = fetch_content(redirect_url, redirect_limit - 1)
+              end
+            rescue => error
+              # Host might no longer exist or some other error, ignore
+            end
+          end
+        end
+      end
+
+      def enabled?
+        nuget_config_exists(config.pwd)
+      end
+
+      def nuget_obj_root
+        config.dig("nuget", "obj_root") || config.pwd
+      end
+
+      def dump_projects?
+        config.dig("nuget", "projects", "dump")
+      end
+
+      def exclude_projects
+        config.dig("nuget", "projects", "exclude") || []
+      end
+
+      def excluded_project?(project_name)
```

Comment:
can you provide a comment here on what the usage here is?  Is this so you can set an exclude for something like `System.*`?  if so, why not specify that in the `ignore.nuget` configuration list instead of creating a new configuration value?

Reply:
In .NET Core, packages are now specified on a per-project basis. That is, there is no packages.config/project.json/etc for the whole solution (repo).

This would make it possible to exclude the packages used by any [test | build-time | non-shipping] projects.
e.g. 
```
projects:
    exclude:
      - "*.Mocks"
      - "*.Tests"
      - Tools.CodeAnalyzers
      - InternalNonShippingService
```

where `Foo.Tests.csproj` is a project that consumes the package `Moq`, which is only used for tests and is never shipped.

I'm happy to take this out if it's against licensed's principles, but for our repository it would result in ~100 packages that would need to be ignored, and make our `.licensed.yml` somewhat unmanageable.

Reply:
just to make sure I understand - the purpose of this is to exclude projects from the repo being evaluated? 

Put another way - In the example above, all of the projects that are intended to match the listed exclusions exist in the repo which is running `licensed`?

If that's the case, you don't need to run licensed from the root of your repo.  licensed allows you to specify multiple evaluation scopes, called [`apps`](https://github.com/github/licensed/blob/master/docs/configuration.md#specifying-multiple-apps) in the documentation.  I would recommend specifying multiple apps for the projects you're looking to enumerate.  You can specify [glob patterns](https://github.com/github/licensed/blob/master/docs/configuration.md#source-path-glob-patterns) for an app source path which will expand to multiple apps when run in a repo.

Does that help your use case?

Reply:
Correct and correct.

I'm not sure it completely does. I think it's possible to implement, but for large solutions it doesn't really seem reasonable to me.

For context:
We don't even look for "source" project files e.g. `Foo.csproj` today; we look directly for `project.assets.json` files which are emitted after `dotnet restore` runs. Mapping directly from `Foo.csproj` -> `project.assets.json` filepath isn't possible without invoking msbuild. Mapping back from a `project.assets.json` -> `Foo.csproj` (or project name `Foo`) is thankfully easy, however.

Which means that we would need to first build up the mapping before being able to find dependencies for a project. Presumably we'd just do this when evaluating the first one and then cache it for the subsequent projects?

In terms of if it would be useful: For our codebase with hundreds of projects (even with most of them having naming conventions), it seems much much safer to exclude rather than include projects. I would honestly have no clue if I successfully included all of our shipping projects using just include patterns.

Reply:
Per our discussion, removed this and at least for now will only support targeting 1 project at a time, where the `source_path` must be the directory containing the `project.assets.json` file.